pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		log.Debugf("updating remote rates with updates: %v", remoteRateUpdates)
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		log.Debugf("updating samplers with remote configuration: %v", samplerconfigPayload)
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		switch configID {
+		case dynamicRatesByService, userDefinedRatesByService:
+			var payload apmsampling.APMSampling
+			_, err := payload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal APMSampling remote config envPayload: %w", err)
+			}
+			remoteRateUpdates = append(remoteRateUpdates, sampler.RemoteRateUpdate{Version: conf.Metadata.Version, Config: payload})
+		case samplerconfig:
+			samplerconfigPayload = &apmsampling.SamplerConfig{}
+			err = json.Unmarshal(conf.Config, samplerconfigPayload)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal SamplerConfig remote config envPayload: %w", err)
+			}
+		}
+	}
+	return remoteRateUpdates, samplerconfigPayload, nil
+}
+
+func (h *RemoteConfigHandler) configPathToID(path string) (apmsamplingRemoteConfigID, error) {
+	pathMatch := h.remoteConfigPathRegex.FindStringSubmatch(path)
+	if pathMatch == nil {
+		return "", fmt.Errorf("failed to match remote config path %s", path)
+	}
+
+	switch pathMatch[1] {
+	case string(dynamicRatesByService):
+		return dynamicRatesByService, nil
+	case string(userDefinedRatesByService):
+		return userDefinedRatesByService, nil
+	case string(samplerconfig):
+		return samplerconfig, nil
+	default:
+		return "", fmt.Errorf("unexpected remote config ID: %s", pathMatch[1])
+	}
+}
+
+func (h *RemoteConfigHandler) updateSamplers(config apmsampling.SamplerConfig) {
+	var confForEnv *apmsampling.SamplerEnvConfig
+	for _, envAndConfig := range config.ByEnv {
+		if envAndConfig.Env == h.agentConfig.DefaultEnv {
+			confForEnv = &envAndConfig.Config
+		}
+	}
+
+	var prioritySamplerTargetTPS float64
```

Comment:
I would
```suggestion
	var prioritySamplerTargetTPS float64 = h.agentConfig.TargetTPS
```

then remove
```
else {
	prioritySamplerTargetTPS = h.agentConfig.TargetTPS
}
```

Will bit easier to read.

Reply:
Andrew suggested the same, I personally prefered to keep if like I did because assignations go from highest to lowest precedence. I'll change it if anyone has a strong opinion of course